File: Python/24-swap-nodes-in-pairs/index2.py
# Definition for singly-linked list.
# class ListNode:
#   def __init__(self, x):
#     self.val = x
#     self.next = None

class Solution:
  def swapPairs(self, head: ListNode) -> ListNode:
    if not head or not head.next:
      return head
    
    new_start = head.next.next
    head, head.next = head.next, head
    head.next.next = self.swapPairs(new_start)
    return head

File: JavaScript/24-swap-nodes-in-pairs/index2.js
/**
 * Definition for singly-linked list.
 * function ListNode(val) {
 *   this.val = val;
 *   this.next = null;
 * }
 */
/**
 * @param {ListNode} head
 * @return {ListNode}
 */
const swapPairs = function (head) {
  if (!head || !head.next) return head;

  const newStart = head.next.next;
  // const temp = head.next;
  // head.next = head;
  // head = temp;

  // 下面解构是赋值可以快速交换两个变量的值
  [head, head.next] = [head.next, head];

  head.next.next = swapPairs(newStart);
  
  return head;
};

